packages/router-core/src/router.ts

```diff
@@ -1440,18 +1440,19 @@ export class RouterCore<
         }
       }
 
-      // If it's already a success, update headers and head content
+      // If the match status is already error, notFound or success, update headers and head content
       // These may get updated again if the match is refreshed
       // due to being stale
-      if (match.status === 'success') {
+      if (['error', 'notFound', 'success'].includes(match.status)) {
+        const { loaderData } = match
         match.headers = route.options.headers?.({
-          loaderData: match.loaderData,
+          ...(loaderData && { loaderData }),
```

Comment:
Why is this bit changing? Same question for the bit down below?

Reply:
yeah seems not necessary.

Reply:
no sorry i was referring to the user facing types